__tests__/types.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * Type only tests. Do not execute.
+ */
+
+import { assert, IsExact, Has } from "conditional-type-checks";
+
+import { oc } from "../src";
+
+/**
+ * Extract type from the array
+ */
+type ArrayType<T> = T extends Array<infer V> ? V : never;
+
+interface X {
+  a?: {
+    b?: string;
+    literal?: "literal";
+    union?: "foo" | "bar";
+    maybeNull: string | null
+    array?: (string | null)[]
+    notNull: string
+  }
+  exists: string
+  getter?: () => string;
+}
+
+declare const x: X;
+
+const resWithDefault = oc(x).a.b("");
+assert<IsExact<typeof resWithDefault, string>>(true);
+
+const resMaybeNull = oc(x).a.maybeNull("")
+assert<IsExact<typeof resMaybeNull, string>>(true);
+

```

Comment:
Can you add a test for the previously supported case of setting default value to `null`:
```typescript
const resMaybeNullWithDefaultNull = oc(x).a.maybeNull(null);
assert<Has<typeof resMaybeNullWithDefaultNull, string>>(true);
assert<Has<typeof resMaybeNullWithDefaultNull, null>>(true);
```

Reply:
Unfortunately it seems to be very hard to support all these cases.

If I revert the implementation to your original the null test passes, but the union test fails:

![image](https://user-images.githubusercontent.com/225712/55594270-a0962600-5747-11e9-8928-9b26a2a7c1f2.png)

Which is in my opinion a lot more important because I can see uses cases where I'd want to remove undefined from a union. It's actually why I created this PR (#19 ). Removing the string from `string | null` union seems really weird to me.

The however the implementation can be modified like this:

![image](https://user-images.githubusercontent.com/225712/55594363-fbc81880-5747-11e9-8747-a249a3c146bd.png)

But it won't still pass completely

![image](https://user-images.githubusercontent.com/225712/55594421-303bd480-5748-11e9-981e-d4b076e6b223.png)

I think the implementation in the PR is the least bad one.

Reply:
Ah right! I ran into this when working on https://github.com/rimeto/ts-transform-optchain. The workaround I found:
```
export interface IDataAccessor<T> {

  /**
   * Data accessor without a default value. If no data exists,
   * `undefined` is returned.
   */
  (noDefaultValue?: undefined): Defined<T> | undefined;

...
```

See if this fixes the problem?

Reply:
Doesn't seem to :/

Reply:
weird -- it appears to work on my checkout of your branch using ts 3.3.3333:
```diff
--- a/src/index.ts
+++ b/src/index.ts
@@ -23,6 +23,11 @@ export type Defined<T> = Exclude<T, undefined>;
 ////////////////////////////
 
 export interface IDataAccessor<T> {
+  /**
+   * Data accessor without a default value. If no data exists,
+   * `undefined` is returned.
+   */
+  (noDefaultValue?: undefined): Defined<T> | undefined;
 
   /**
    * Data accessor with default value.
@@ -34,7 +39,7 @@ export interface IDataAccessor<T> {
    * Data accessor without a default value. If no data exists,
    * `undefined` is returned.
    */
-  (): T | undefined
+  (nullDefaultValue: T extends null ? null : never): Defined<T>;
 }
 
 ///////////////////////////
```

Reply:
Ah missunderstood your patch previously. All green now! 🎉